### PR TITLE
Error instead of `panic!` when running Ruff from a deleted directory (#16903)

### LINKED ISSUE #16903: [Panic] ruff crash when executed from a deleted folder on mac
How to reproduce : 

in terminal 1 : 

``
mkdir tmp
cd tmp
``

in terminal 2

```
rm -rf tmp
```

back in terminal 1 : 

```
RUST_BACKTRACE=full ruff check

error: Ruff crashed. If you could open an issue at:

    https://github.com/astral-sh/ruff/issues/new?title=%5BPanic%5D

...quoting the executed command, along with the relevant file contents and `pyproject.toml` settings, we'd be very appreciative!

thread 'main' panicked at /Users/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/path-dedot-3.1.1/src/lib.rs:330:70:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }
stack backtrace:
   0:        0x10f6a86e8 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hcaf66bc4c0c453df
   1:        0x10e7e360b - core::fmt::write::hc9c5f1836b413410
   2:        0x10f6a3f12 - std::io::Write::write_fmt::h49df280499063c09
   3:        0x10f6a99da - std::panicking::default_hook::{{closure}}::h52c0b2f44f6107c5
   4:        0x10f6a9642 - std::panicking::default_hook::h5a6cf31501c161b2
   5:        0x10ecd794a - ruff::run::{{closure}}::hd41b263f42227ea1
   6:        0x10f6aab9e - std::panicking::rust_panic_with_hook::hda4640ee332466e9
   7:        0x10f6a9ff5 - std::panicking::begin_panic_handler::{{closure}}::haa3060694b34ea3d
   8:        0x10f6a8bc9 - std::sys::backtrace::__rust_end_short_backtrace::h8eb44913cfe71457
   9:        0x10f6a9c3c - _rust_begin_unwind
  10:        0x10f750d4a - core::panicking::panic_fmt::h31edc3d6ff0aadca
  11:        0x10f7511c5 - core::result::unwrap_failed::h2169d1cb8033298b
  12:        0x10e9b57c9 - core::ops::function::FnOnce::call_once::h4a1f9b6c37c4efb9
  13:        0x10ec9e540 - once_cell::imp::OnceCell<T>::initialize::{{closure}}::hd57f3dc11ebdb7bf
  14:        0x10e9b392a - once_cell::imp::initialize_or_wait::h1b6a1b2a2c5ea054
  15:        0x10f776c3c - once_cell::imp::OnceCell<T>::initialize::h6c5acec7735a0040
  16:        0x10ebcd23b - ruff::resolve::resolve::h3abf4f7e3ef26cd9
  17:        0x10ecd7ad1 - ruff::check::h93bfe06211e45642
  18:        0x10ecd6d01 - ruff::run::h4a796964f3396cea
  19:        0x10ecdc441 - ruff::main::h12661c1f2feb882f
  20:        0x10ece1936 - std::sys::backtrace::__rust_begin_short_backtrace::h61cdb7c3f5f8eba8
  21:        0x10ecdfb0c - std::rt::lang_start::{{closure}}::hf5a90ee77f97c307
  22:        0x10f69b27f - std::rt::lang_start_internal::hb1473645dbe40065
  23:        0x10ecdc97c - _main
  24:     0x7ff80d741345 - <unknown>
```

## Context

* Python 3.12.4
* MacOS Sonoma 14.6

### PR DESCRIPTION
**Pull request by maxmynter**
Closes #16903

## Summary
Check if the current working directory exist. If not, provide an error instead of panicking.

Fixed a stale comment in `resolve_default_files`.

## Test Plan
I added a test to the `resolve_files.rs`. 

Manual testing follows steps of #16903 :
- Terminal 1
```bash
mkdir tmp
cd tmp
```
- Terminal 2
```bash
rm -rf tmp
```
- Terminal 1
```bash
ruff check
```

## Open Issues / Questions to Reviewer
All tests pass when executed with `cargo nextest run`.

However, with `cargo test` the parallelization makes the other tests fail as we change the `pwd`.

Serial execution with `cargo test` seems to require [another dependency or some workarounds](https://stackoverflow.com/questions/51694017/how-can-i-avoid-running-some-tests-in-parallel).

Do you think an additional dependency or test complexity is worth testing this small edge case, do you have another implementation idea, or should i rather remove the test?
  
---
P.S.: I'm currently participating in a batch at the [Recurse Center](https://www.recurse.com/) and would love to contribute more for the next six weeks to improve my Rust. Let me know if you're open to mentoring/reviewing and/or if you have specific areas where help would be most valued.

### CODE REVIEW

File: crates/ruff/tests/resolve_files.rs [hunk 1]
```diff
@@ -101,3 +101,28 @@ fn check_project_from_project_subdirectory_respects_includes() {
...
+fn check_in_deleted_directory_errors() {
+    let temp_dir = tempfile::tempdir().unwrap();
+    let temp_path = temp_dir.path().to_path_buf();
+    let original_dir = std::env::current_dir().unwrap();
+    std::env::set_current_dir(&temp_path).unwrap();
+    std::mem::drop(temp_dir);
+
+    insta::with_settings!({
+        filters => TEST_FILTERS.to_vec()
+    },
```
Comment (ntBre) [inline crates/ruff/tests/resolve_files.rs:115]: I don't think you actually need these filters and could leave them out. Paths only cause problems when they appear in the snapshot output.

**Review by ntBre — COMMENTED** [review]
Thanks for working on this! I took a look at this earlier in the week, and I was wondering how to go about testing it too.

I think if you move the new integration test to its own file, it will be built and run separately by `cargo test` and avoid the parallelization issues you ran into. Then you can also skip saving the `original_dir` and restoring it since the test should be run on its own.

Alternatively, we could possibly skip testing this to avoid changing the cwd at all, which feels a little scary. What do you think @MichaReiser?

File: crates/ruff/tests/resolve_files.rs [hunk 2]
```diff
@@ -101,3 +101,28 @@ fn check_project_from_project_subdirectory_respects_includes() {
...
     });
 }
+
+#[test]
+fn check_in_deleted_directory_errors() {
+    let temp_dir = tempfile::tempdir().unwrap();
+    let temp_path = temp_dir.path().to_path_buf();
+    let original_dir = std::env::current_dir().unwrap();
+    std::env::set_current_dir(&temp_path).unwrap();
+    std::mem::drop(temp_dir);
```
[the remark this answers is not in the document]
Comment (maxmynter) [reply crates/ruff/tests/resolve_files.rs:111]: Agree that it reads nicer. 
I will also just import from `std::env` so we can use directory operations without qualifiers.

File: crates/ruff/src/resolve.rs [hunk 3]
```diff
@@ -21,6 +21,10 @@ pub fn resolve(
     config_arguments: &ConfigArguments,
     stdin_filename: Option<&Path>,
 ) -> Result<PyprojectConfig> {
+    // Ensure directory exists before proceeding
+    if std::env::current_dir().is_err() {
+        return Err(anyhow!("Working directory does not exist"));
```
[the remark this answers is not in the document]
Comment (maxmynter) [reply crates/ruff/src/resolve.rs:26]: Agree on your proposal for rewriting. It makes the intentions of the check clear and we can get rid of the comment.

`path_dedot::CWD` is used a couple of times throughout the function so I will keep it outside of the if clause.

**Comment by maxmynter** [comment]
Having the test in a dedicated file now works for both `cargo test` and `cargo nextest`.

File: crates/ruff/src/resolve.rs [hunk 4]
```diff
@@ -21,6 +21,9 @@ pub fn resolve(
     config_arguments: &ConfigArguments,
     stdin_filename: Option<&Path>,
 ) -> Result<PyprojectConfig> {
+    let Ok(_cwd) = std::env::current_dir() else {
```
Comment (MichaReiser) [inline crates/ruff/src/resolve.rs:24]: It feels off that the path isn't used here. I think we should change the code to call `cwd.parse_dot()` and use it instead of `&path_dedot::CWD` below

File: crates/ruff/tests/direxist_guard.rs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+const BIN_NAME: &str = "ruff";
+
+#[test]
+fn check_in_deleted_directory_errors() {
+    let temp_dir = tempfile::tempdir().unwrap();
+    let temp_path = temp_dir.path().to_path_buf();
+    set_current_dir(&temp_path).unwrap();
+    drop(temp_dir);
+
+    assert_cmd_snapshot!(Command::new(get_cargo_bin(BIN_NAME)).arg("check"), @r###"
```
Comment (MichaReiser) [inline crates/ruff/tests/direxist_guard.rs:16]: I agree with @ntBre that changing the working directory is problematic. Let's only override the current working directory for the spanned CLI command.

```suggestion
    assert_cmd_snapshot!(Command::new(get_cargo_bin(BIN_NAME)).arg("check").current_dir(&temp_path), @r###"
```

Comment (ntBre) [reply crates/ruff/tests/direxist_guard.rs:16]: I think I tried `current_dir` when I looked at this, and it caused a different panic in the `Command` invocation, separate from the one we're trying to test. But it would be good to double check, and maybe I missed something.

Comment (maxmynter) [reply crates/ruff/tests/direxist_guard.rs:16]: Yes, the `assert_cmd_snapshot` or `current_dir` calls an `unwrap()` somewhere. The test fails with: 

```bash
[...] called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory"
```

Comment (MichaReiser) [reply crates/ruff/tests/direxist_guard.rs:16]: Hmm I see. Let's add some module level documentation (or at least test level) explaining why this test *has to be its own module*.

File: crates/ruff/src/resolve.rs [hunk 6]
```diff
@@ -58,11 +61,9 @@ pub fn resolve(
     // that directory. (With `Strategy::Hierarchical`, we'll end up finding
     // the "closest" `pyproject.toml` file for every Python file later on,
     // so these act as the "default" settings.)
-    if let Some(pyproject) = pyproject::find_settings_toml(
-        stdin_filename
-            .as_ref()
-            .unwrap_or(&path_dedot::CWD.as_path()),
-    )? {
+    if let Some(pyproject) =
+        pyproject::find_settings_toml(stdin_filename.as_ref().unwrap_or(&cwd.as_path()))?
```
Comment (MichaReiser) [inline crates/ruff/src/resolve.rs:65]: Is it intentional that you're only using `parse_dot` in some paths? 

I think I'd change the method to 
```rust
    let Ok(cwd) = std::env::current_dir() else {
        bail!("Working directory does not exist")
    };
		let cwd = cwd.parse_dot()?;
```

Comment (ntBre) [reply crates/ruff/src/resolve.rs]: That looks right to me. I *think* we could skip calling `parse_dot` at all since `path_dedot::CWD` doesn't:

https://github.com/magiclen/path-dedot/blob/36c197f922d4c8779594ce25b8ffc6cb0315c6a8/src/lib.rs#L329-L330

but I think Micha's point was that we should be consistent one way or the other.

File: crates/ruff/src/resolve.rs [hunk 7]
```diff
@@ -58,11 +63,9 @@ pub fn resolve(
     // that directory. (With `Strategy::Hierarchical`, we'll end up finding
     // the "closest" `pyproject.toml` file for every Python file later on,
     // so these act as the "default" settings.)
-    if let Some(pyproject) = pyproject::find_settings_toml(
-        stdin_filename
-            .as_ref()
-            .unwrap_or(&path_dedot::CWD.as_path()),
-    )? {
+    if let Some(pyproject) =
+        pyproject::find_settings_toml(stdin_filename.as_ref().unwrap_or(&cwd.as_ref()))?
```
Comment (ntBre) [inline crates/ruff/src/resolve.rs:67]: nit: You do need `as_ref` currently, but `stdin_filename` is already an `Option<&Path>`, so you can actually simplify both of these cases (here and below on line 137) to this:

```suggestion
        pyproject::find_settings_toml(stdin_filename.unwrap_or(&cwd))?
```

File: crates/ruff/tests/direxist_guard.rs [hunk 8]
```diff
@@ -0,0 +1,30 @@
+//! Test to verify Ruff's behavior when run from deleted directory.
+//! It has to be isolated in a separate module.
+//! Tests in the same module become flaky under `cargo test`s parallel execution
+//! due to in-test working directory manipulation.
+
+#![cfg(not(target_family = "wasm"))]
```
Comment (ntBre) [inline crates/ruff/tests/direxist_guard.rs:6]: I think the Windows CI failure is real. I think Windows will keep the directory around as long as a process is using it. Maybe we need to exclude both `wasm` and `windows` targets? I always forget the syntax for this, but I think it's something like this:

```suggestion
#![cfg(not(any(target_family = "wasm", target_family = "windows")))]
```

**Review by ntBre — COMMENTED** [review]
Thanks for following up on this, it's looking good. Just one more nit to avoid `as_ref` calls and an idea for fixing Windows CI.

I'll defer to Micha for the final approval, but this looks good to me once CI passes.

File: crates/ruff/src/resolve.rs (hunk 6 above)
Comment (maxmynter) [reply crates/ruff/src/resolve.rs]: Interesting. I used it because the first line on the `path_dedot` crate docs say it's purpose is to handle filepaths that contain dots. 

But if we only use(d) it as a global variable to access the cwd, I think we can drop it.

Addressed in [98c7711](https://github.com/astral-sh/ruff/pull/17054/commits/98c7711395a68da11b4d93c7c71e6fc0e84f182f)

File: crates/ruff/tests/direxist_guard.rs (hunk 8 above)
Comment (maxmynter) [reply crates/ruff/tests/direxist_guard.rs:6]: I'll leave the Windows CI failures to you, if that's fine... I don't have access to a Windows machine.
